tests/unit/core/mender-artifact_test.cpp

```diff
@@ -0,0 +1,229 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include <filesystem>
+#include <fstream>
+
+extern "C" {
+#include "mender-artifact.h"
+#include "mender-deployment-data.h"
+#include "mender-update-module.h"
+}
+
+using namespace std;
+namespace fs = filesystem;
+
+class MenderArtifactTest : public ::testing::Test {
+protected:
+    string   artifact_data;
+    void    *data;
+    fs::path artifact_path;
+    fs::path script_file_name;
+
+    string CreateArtifact() {
+        string script = R"(#! /bin/sh
+        DIRNAME=$(dirname $0)
+        echo foobar > ${DIRNAME}/testdata
+        mender-artifact \
+        write module-image \
+        --compression none \
+        --type unit-test \
+        -n unit-test-artifact \
+        --device-type unit-test  \
+        -f ${DIRNAME}/testdata \
+        -o ${DIRNAME}/unit-test-artifact.mender || exit 1)";
+
+        fs::path tmp_dir = fs::temp_directory_path();
+        script_file_name = tmp_dir / "test-script.sh";
+
+        ofstream os(script_file_name.c_str(), ios::out);
+
+        os << script;
+        os.close();
+
+        chmod(script_file_name.c_str(), S_IRUSR | S_IWUSR | S_IXUSR);
+        system(script_file_name.c_str());
+
+        artifact_path = tmp_dir / "unit-test-artifact.mender";
+        ifstream file(artifact_path);
+        return string((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
+    }
+
+    void create_download_data(mender_update_module_t          *update_module,
+                              mender_deployment_data_t        *deployment_data,
+                              mender_artifact_download_data_t *mock_download_data) {
+        mender_create_deployment_data("id", "unit-test-artifact", &deployment_data);
+
+        update_module->artifact_type = "unit-test";
+        update_module->callbacks[0]
+            = [](MENDER_ARG_UNUSED mender_update_state_t state, MENDER_ARG_UNUSED mender_update_state_data_u data) { return MENDER_OK; };
+        mender_update_module_register(update_module);
+
+        mock_download_data->deployment    = deployment_data;
+        mock_download_data->update_module = update_module;
+    }
+
+    mender_update_module_t          *update_module;
+    mender_deployment_data_t        *deployment_data;
+    mender_artifact_download_data_t *mock_download_data;
+
+    void SetUp() {
+        update_module      = (mender_update_module_t *)malloc(sizeof(mender_update_module_t));
+        deployment_data    = (mender_deployment_data_t *)malloc(sizeof(mender_deployment_data_t));
+        mock_download_data = (mender_artifact_download_data_t *)malloc(sizeof(mender_artifact_download_data_t));
```

Comment:
On first read I thought that it didn't make sense to use dynamic memory for these on the tests, and we should just use C++ (implicit) constructor/destructor on these `struct`. But then I saw that `mender_update_module_unregister_all` does free the `update_module`... From the API point of vies, is it really correct that `mender_update_module_unregister_all` takes ownership of the `update_module`?

Reply:
As the code is today, yes: it is the right way. I won't insist this late in the API design, it was just thinking out loud.